doc/man3/OSSL_PARAM.pod

```diff
@@ -108,20 +108,17 @@ The C<data_type> field can be one of the following types:
 
 =over 4
 
-=item C<OSSL_PARAM_INTEGER>
+=item C<OSSL_PARAM_INT>, C<OSSL_PARAM_UINT> C<OSSL_PARAM_INT>, C<OSSL_PARAM_UINT>, C<OSSL_PARAM_INT64>, C<OSSL_PARAM_UINT64>, C<OSSL_PARAM_LONG>, C<OSSL_PARAM_ULONG>, C<OSSL_PARAM_SIZE_T>
 
-=item C<OSSL_PARAM_UNSIGNED_INTEGER>
+The parameter is an integer that has the C type of C<int>, C<unsigned
+int>, C<int64_t>, C<uint64_t>, C<long>, C<unsigned long>, or C<size_t>
+respectively and is represented in its native form.
 
-The parameter data is an integer (signed or unsigned) of arbitrary
-length, organized in native form, i.e. most significant byte first on
-Big-Endian systems, and least significant byte first on Little-Endian
-systems.
+=item C<BIGNUM>
 
-=item C<OSSL_PARAM_REAL>
-
-=for comment It's still debated if we need this or not.
-
-The parameter data is a floating point value in native form.
+The parameter is an OpenSSL C<BIGNUM> and is represented using
+BN_lebin2bn() or BN_bin2bn(), depending on whether the platform is
+little-endian or big-endian.
```

Comment:
So it's not an OpenSSL BIGNUM, it's big number organized in native byte order. This could just as well be an export of a gmp number, done with [`mpz_export`](https://gmplib.org/manual/Integer-Import-and-Export.html)

Reply:
I don't know the details of that the bin2bn/bn2bin functions do.  For example, do they preserve sign?  More importantly, I don't have to know.

Reply:
They don't preserve sign (incidently, neither does mpz_export).
(that was actually news to me)

Reply:
Yes, it was news to me as well.  I think we need a better serialization format that preserves the sign.  Or just pass the BIGNUM directly.  I am inclined to prefer the latter.  Thoughts?

Reply:
I think we have to have a clear format and not just say "whatever the format of BIGNUM bn2bin is" and definitely not a BIGNUM directly. That format of bn2bin is (presumably) defined somewhere so this shouldn't be hard. We don't want to require Provider authors to have to link to libcrypto, so if they want to use gmp instead, then that's fine. Yes it should preserve the sign.

Reply:
@mattcaswell, I'll have you note this from the [GMP manual](https://gmplib.org/manual/Integer-Import-and-Export.html#Integer-Import-and-Export):

> The sign of *op* is ignored, just the absolute value is exported. An application can use `mpz_sgn` to get the sign and handle it as desired.

So we're definitely not alone in this.

Reply:
Hmm...thinking again about the sign thing. Perhaps we don't need it. I can't think of a case where any crypto parameter can ever validly be negative.

Reply:
Yeah I was thinking something similar, that we probably can get away with it.

Reply:
I think that only allowing non-negative BN's is okay for now, I can't think of a reason for them at the moment.

We don't know what will be invented in the future, so it is possible they will be required eventually.  I think it is safe to defer supporting them until such an algorithm does exist.  Nothing before then can want them, so there shouldn't be a backwards compatibility problem.